src/Roslyn.Diagnostics.Analyzers/Core/DoNotCopyValue.cs

```diff
@@ -547,14 +549,36 @@ public override void VisitForEachLoop(IForEachLoopOperation operation)
 
                 var instance = operation.Collection;
                 var instance2 = (operation.Collection as IConversionOperation)?.Operand;
-                if (Acquire(operation.Collection) != RefKind.Ref)
+
+                switch (Acquire(operation.Collection))
                 {
-                    instance = null;
-                    instance2 = null;
+                    case RefKind.Ref:
+                        // No special requirements
+                        break;
+
+                    case RefKind.RefReadOnly when operation.Syntax is CommonForEachStatementSyntax syntax && operation.SemanticModel.GetForEachStatementInfo(syntax).GetEnumeratorMethod is { IsReadOnly: true }:
```

Comment:
This is getting too difficult to manage - IMO this analyzer is not sustainable unless rewritten as a flow-based analyzer. I'll let you decide how to take this forward.

Reply:
> IMO this analyzer is not sustainable unless rewritten as a flow-based analyzer.

Flow based analysis does not help here. Every language construct needs to be allow-listed as it is encountered.